flink-end-to-end-tests/flink-end-to-end-tests-common/src/main/java/org/apache/flink/tests/util/flink/container/FlinkContainerBuilder.java

```diff
@@ -0,0 +1,248 @@
+    public FlinkContainerBuilder enableHAService() {
+        this.enableHAService = true;
+        return this;
+    }
+
+    /** Builds {@link FlinkContainer}. */
+    public FlinkContainer build() {
+        GenericContainer<?> haService = null;
+        if (enableHAService) {
+            enableHAServiceConfigurations();
+            haService = buildZookeeperContainer();
+        }
+
+        this.conf.set(JobManagerOptions.ADDRESS, FlinkContainer.JOB_MANAGER_HOSTNAME);
+        this.conf.set(
+                CheckpointingOptions.CHECKPOINTS_DIRECTORY,
+                FlinkContainer.CHECKPOINT_PATH.toAbsolutePath().toUri().toString());
+
+        final GenericContainer<?> jobManager = buildJobManagerContainer();
+        final List<GenericContainer<?>> taskManagers = buildTaskManagerContainers();
+
+        if (enableHAService) {
+            try {
+                Path recoveryPath = Files.createTempDirectory("flink-recovery");
+                jobManager.withFileSystemBind(
+                        recoveryPath.toAbsolutePath().toString(),
+                        FlinkContainer.HA_STORAGE_PATH.toAbsolutePath().toString());
+            } catch (IOException e) {
+                throw new IllegalStateException("Failed to create temporary recovery directory", e);
+            }
+        }
+
+        try {
+            Path checkpointPath = Files.createTempDirectory("flink-checkpoint");
+            jobManager.withFileSystemBind(
+                    checkpointPath.toAbsolutePath().toString(),
+                    FlinkContainer.CHECKPOINT_PATH.toAbsolutePath().toString());
+        } catch (IOException e) {
+            throw new IllegalStateException("Failed to create temporary checkpoint directory", e);
+        }
+
+        return new FlinkContainer(jobManager, taskManagers, haService, conf);
+    }
+
+    // --------------------------- Helper Functions -------------------------------------
+
+    private GenericContainer<?> buildJobManagerContainer() {
+        // Configure JobManager
+        final Configuration jobManagerConf = new Configuration();
+        jobManagerConf.addAll(this.conf);
+        // Build JobManager container
+        final ImageFromDockerfile jobManagerImage;
+        try {
+            jobManagerImage =
+                    new FlinkImageBuilder().setConfiguration(jobManagerConf).asJobManager().build();
+        } catch (Exception e) {
+            throw new RuntimeException("Failed to build JobManager image", e);
+        }
+        final GenericContainer<?> jobManager = buildContainer(jobManagerImage);
+        // Setup network for JobManager
+        jobManager
+                .withNetworkAliases(FlinkContainer.JOB_MANAGER_HOSTNAME)
+                .withExposedPorts(jobManagerConf.get(RestOptions.PORT));
+        // Setup logger
+        if (this.logger != null) {
+            jobManager.withLogConsumer(new Slf4jLogConsumer(this.logger).withPrefix("JobManager"));
```

Comment:
I think syncing with logger's level is not quite useful in this case. Instead I added a method in `FlinkContainerBuilder` for configuring log4j properties directly. By default it will use `log4j-console.properties` under flink-dist, with root logger  level set to INFO, and properties specified in builder will append or overwrite entries in `log4j-console.properties`.